File: crates/flighthq-scene-gl/src/vertex_color_gl_mesh_material_renderer.rs
//! The built-in VertexColor forward renderer.
//!
//! Ports `@flighthq/scene-gl` `vertexColorGlMeshMaterialRenderer.ts`. Lighting-
//! independent: the unlit shader's VERTEX_COLOR variant multiplies the mesh's
//! interpolated `color0` attribute by the linear tint. `bind` selects the variant,
//! uploads the camera view-projection and the linear tint, and defaults the
//! `color0` generic attribute to opaque white so a mesh WITHOUT a `color0`
//! attribute renders the tint alone. Lights are ignored.

use flighthq_materials::unlit_materials::vertex_color_material_kind;
use flighthq_render_gl::GlRenderState;
use flighthq_types::camera::Camera;
use flighthq_types::pbr_material::MaterialAlphaMode;
use flighthq_types::scene_render::{SceneLightBlock, SceneRenderProxy};
use glow::HasContext;

use crate::gl_mesh_material_registry::{
    GlMeshMaterialRenderer, MeshMaterial, register_gl_mesh_material_renderer,
};
use crate::gl_mesh_program::{
    begin_gl_mesh_draw, draw_gl_mesh_subset, set_gl_mesh_view_projection,
};
use crate::gl_pbr_standard_bind::unpack_color_to_linear;
use crate::gl_scene_runtime::{GlMeshUpload, GlSceneRuntime};
use crate::gl_unlit_prelude::{GlUnlitDefineKey, bind_gl_unlit_surface, ensure_gl_unlit_program};

/// The built-in VertexColor renderer.
pub struct VertexColorGlMeshMaterialRenderer;

impl GlMeshMaterialRenderer for VertexColorGlMeshMaterialRenderer {
    fn bind(
        &self,
        state: &mut GlRenderState,
        scene: &mut GlSceneRuntime,
        material: Option<&dyn MeshMaterial>,
        _lights: &SceneLightBlock,
        camera: &Camera,
    ) {
        let vertex_color = material.and_then(|m| m.as_vertex_color());
        let key = GlUnlitDefineKey {
            alpha_mask_enabled: vertex_color
                .is_some_and(|v| v.alpha_mode == MaterialAlphaMode::Mask),
            has_color_map: false,
            vertex_color: true,
        };
        let program = ensure_gl_unlit_program(state, scene, &key);
        begin_gl_mesh_draw(
            state,
            scene,
            &program.base,
            vertex_color.is_some_and(|v| v.double_sided),
        );
        set_gl_mesh_view_projection(&state.gl, program.base.loc_view_projection.as_ref(), camera);

        // Default the color0 generic vertex attribute to opaque white so a mesh
        // WITHOUT a color0 attribute renders the tint alone (matching the wgpu
        // path) instead of multiplying by the (0,0,0,1) default, which would
        // render black. A mesh that DOES carry color0 enables the array and
        // overrides this.
        unsafe {
            state.gl.vertex_attrib_4_f32(4, 1.0, 1.0, 1.0, 1.0);
        }

        let (color, alpha_cutoff) = match vertex_color {
            None => ([1.0, 1.0, 1.0, 1.0], 0.5),
            Some(v) => (unpack_color_to_linear(v.tint), v.alpha_cutoff),
        };
        bind_gl_unlit_surface(state, &program, &color, 1.0, alpha_cutoff);
    }

    fn draw(
        &self,
        state: &mut GlRenderState,
        scene: &mut GlSceneRuntime,
        proxy: &SceneRenderProxy,
        upload: &GlMeshUpload,
    ) {
        draw_gl_mesh_subset(state, scene, proxy, upload);
    }
}

/// Registers the built-in VertexColor renderer for `VertexColorMaterialKind`.
pub fn register_vertex_color_gl_material(scene: &mut GlSceneRuntime) {
    register_gl_mesh_material_renderer(
        scene,
        vertex_color_material_kind(),
        Box::new(VertexColorGlMeshMaterialRenderer),
    );
}

#[cfg(test)]
mod tests {
    use super::*;
    use flighthq_materials::unlit_materials::vertex_color_material_kind;

    use crate::gl_mesh_material_registry::get_gl_mesh_material_renderer;
    use crate::gl_scene_runtime::create_gl_scene_runtime;

    // register_vertex_color_gl_material

    #[test]
    fn register_vertex_color_gl_material_registers_for_the_vertex_color_kind() {
        let mut scene = create_gl_scene_runtime();
        register_vertex_color_gl_material(&mut scene);
        assert!(get_gl_mesh_material_renderer(&scene, vertex_color_material_kind()).is_some());
    }
}

File: packages/scene-gl/src/vertexColorGlMeshMaterialRenderer.ts
import type { LinearColor } from '@flighthq/color';
import { unpackColorToLinear } from '@flighthq/color';
import type {
  Camera,
  GlMeshMaterialRenderer,
  GlRenderState,
  Material,
  MeshGeometry,
  SceneLightBlock,
  SceneRenderProxy,
  VertexColorMaterial,
} from '@flighthq/types';
import { VertexColorMaterialKind } from '@flighthq/types';

import { registerGlMeshMaterialRenderer } from './glMeshMaterialRegistry';
import { beginGlMeshDraw, drawGlMeshSubset, setGlMeshViewProjection } from './glMeshProgram';
import { getGlSceneRuntime } from './glSceneRuntime';
import type { GlUnlitDefineKey } from './glUnlitPrelude';
import { bindGlUnlitSurface, ensureGlUnlitProgram } from './glUnlitPrelude';

// The built-in VertexColor forward renderer (GlMeshMaterialRenderer for VertexColorMaterialKind).
// Lighting-independent: the unlit shader's VERTEX_COLOR variant multiplies the mesh's interpolated
// color0 attribute by the linear tint. Geometry without a color0 attribute leaves the attribute
// unbound (its default), so a tinted black surface results — author color0 (or import it via glTF)
// to drive this material. Lights are ignored. See registerVertexColorGlMaterial to install it.
export const vertexColorGlMeshMaterialRenderer: GlMeshMaterialRenderer = {
  bind(
    state: GlRenderState,
    material: Readonly<Material> | null,
    _lights: Readonly<SceneLightBlock>,
    camera: Readonly<Camera>,
  ): void {
    const gl = state.gl;
    const vertexColor = material as Readonly<VertexColorMaterial> | null;
    const program = ensureGlUnlitProgram(state, defineKeyForMaterial(vertexColor));
    beginGlMeshDraw(state, program, vertexColor !== null && vertexColor.doubleSided);
    setGlMeshViewProjection(gl, program.locViewProjection, camera);

    // Default the color0 generic vertex attribute to opaque white so a mesh WITHOUT a color0 attribute
    // renders the tint alone (matching the wgpu path) instead of multiplying by the (0,0,0,1) default,
    // which would render black. A mesh that DOES carry color0 enables the array and overrides this.
    gl.vertexAttrib4f(4, 1, 1, 1, 1);

    if (vertexColor === null) {
      bindGlUnlitSurface(state, program, WHITE, 1, null, 0.5);
      return;
    }
    unpackColorToLinear(scratchRgba, vertexColor.tint);
    bindGlUnlitSurface(state, program, scratchRgba, 1, null, vertexColor.alphaCutoff);
  },

  draw(state: GlRenderState, proxy: Readonly<SceneRenderProxy>, geometry: Readonly<MeshGeometry>): void {
    const program = getGlSceneRuntime(state).activeMeshProgram;
    if (program === null) return;
    drawGlMeshSubset(state, program, proxy, geometry);
  },
};

// Registers the built-in VertexColor renderer for VertexColorMaterialKind on this state. Opt-in (no
// top-level side effect); call once per GlRenderState before drawScene so meshes with
// VertexColorMaterials draw.
export function registerVertexColorGlMaterial(state: GlRenderState): void {
  registerGlMeshMaterialRenderer(state, VertexColorMaterialKind, vertexColorGlMeshMaterialRenderer);
}

function defineKeyForMaterial(material: Readonly<VertexColorMaterial> | null): GlUnlitDefineKey {
  return {
    alphaMaskEnabled: material !== null && material.alphaMode === 'mask',
    hasColorMap: false,
    // VertexColor samples no map, so the uv transform never applies.
    hasUvTransform: false,
    vertexColor: true,
  };
}

const scratchRgba: LinearColor = [0, 0, 0, 0];
const WHITE: LinearColor = [1, 1, 1, 1];
